Modernize base_uint:

HowardHinnant opened the pull request:
*  Add construction and assignment from a generic
   contiguous container.  Both compile-time and run time
   safety checks are made to ensure the safety of this
   conversion.

*  Remove base_uint::copyFrom.  The generic copy assignment
   operator now does this functionality with enhanced
   safety and better syntax.

*  Remove construction from and dedendence on Blob.
   The generic constructor and assignment now handle this
   functionality.

*  Fix client code to adhere to this new API.

*  Removed the use of fromVoid in PeerImp.cpp as it was
   an inappropriate use of this dangerous API.  The
   generic container constructors do it with enhanced
   safety and better syntax.

scottschurr commented on src/ripple/app/ledger/OrderBookDB.cpp:
```diff
@@ -114,14 +114,10 @@ void OrderBookDB::update(
-                    sfTakerGetsIssuer));
-                book.out.currency.copyFrom (sle->getFieldH160(
-                    sfTakerGetsCurrency));
+                book.in.currency = sle->getFieldH160(sfTakerPaysCurrency);
```
Really nice!

scottschurr commented on src/ripple/protocol/STAccount.h:
```diff
@@ -34,7 +34,7 @@ class STAccount final
     // overhead in a ripple::uint160.  However, so the serialized format of the
     // STAccount stays unchanged, we serialize and deserialize like an STBlob.
-    uint160 value_;
+    AccountID value_;
```
:+1:

scottschurr commented on src/ripple/basics/base_uint.h:
```diff
@@ -126,25 +126,45 @@ class base_uint
+        return *this;
     }
 
     /* Construct from a raw pointer.
```
Bummer.  I really wish we could remove `fromVoid()`.  I looked at it a bit.  There are 4 uses of `fromVoid()` left in the code base.  Two of them are in unit tests and could easily be fixed.  Unfortunately the other two are harder to fix.  They are:

-  https://github.com/ripple/rippled/blob/develop/src/ripple/nodestore/backend/MemoryFactory.cpp#L127
- https://github.com/ripple/rippled/blob/develop/src/ripple/nodestore/impl/DecodedBlob.cpp#L85

For these two, the available information really is a `void*`.  Those two will have to be fixed before `fromVoid()` can be removed.  I'm uncertain how complicated those fixes will be, but they're not trivial local changes.

HowardHinnant replied:
I had overlooked the `fromVoid()` in tests.  Fixing those, thanks.

HowardHinnant commented:
I don't feel strongly either way myself.  You make a good point about testing it.

ripplelabs-jenkins commented:
## Jenkins Build Summary

Built from [this commit](https://github.com/ripple/rippled/commit/6fec78cf6c14448713014c76e64b10b19f684d49)

Built at __20190524 - 23:01:03__

### Test Results

Build Type | Log | Result | Status
---------- | --- | ------ | ------
rpm | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2951/1//artifact/rpm.txt) | 1178 cases, 0 failed, t: n/a | PASS :white_check_mark: 
msvc.Debug | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2951/1//artifact/msvc.Debug.txt) | 1178 cases, 0 failed, t: 566s | PASS :white_check_mark: 
msvc.Debug, <br/>NINJA_BUILD=true | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2951/1//artifact/msvc.Debug_NINJA_BUILD_true.txt) | 1178 cases, 0 failed, t: 600s | PASS :white_check_mark: 
msvc.Debug <br/>-Dunity=OFF | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2951/1//artifact/msvc.Debug_Dunity_OFF.txt) | 1177 cases, 0 failed, t: 1087s | PASS :white_check_mark: 
msvc.Release | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2951/1//artifact/msvc.Release.txt) | 1178 cases, 0 failed, t: 402s | PASS :white_check_mark: 
gcc.Release <br/>-Dassert=ON, <br/>MANUAL_TESTS=true | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2951/1//artifact/gcc.Release_Dassert_ON_MANUAL_TESTS_true.txt) | 915 cases, 0 failed, t: 4m48s | PASS :white_check_mark: 
docs, <br/>TARGET=docs | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2951/1//artifact/docs_TARGET_docs.txt) | 1 cases, 0 failed, t: 0m1s | PASS :white_check_mark: 
clang.Debug | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2951/1//artifact/clang.Debug.txt) | 1178 cases, 0 failed, t: 2m43s | PASS :white_check_mark: 
gcc.Debug <br/>-Dcoverage=ON, <br/>TARGET=coverage_report, <br/>SKIP_TESTS=true | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2951/1//artifact/gcc.Debug_Dcoverage_ON_TARGET_coverage_report_SKIP_TESTS_true.txt) | 1178 cases, 0 failed, t: 16m37s | PASS :white_check_mark: 
gcc.Debug | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2951/1//artifact/gcc.Debug.txt) | 1178 cases, 0 failed, t: 2m53s | PASS :white_check_mark: 
clang.Debug <br/>-Dunity=OFF | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2951/1//artifact/clang.Debug_Dunity_OFF.txt) | 1177 cases, 0 failed, t: 9m49s | PASS :white_check_mark: 
clang.Release <br/>-Dassert=ON | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2951/1//artifact/clang.Release_Dassert_ON.txt) | 1178 cases, 0 failed, t: 4m44s | PASS :white_check_mark: 
gcc.Debug <br/>-Dunity=OFF | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2951/1//artifact/gcc.Debug_Dunity_OFF.txt) | 1177 cases, 0 failed, t: 10m54s | PASS :white_check_mark: 
gcc.Release <br/>-Dassert=ON | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2951/1//artifact/gcc.Release_Dassert_ON.txt) | 1178 cases, 0 failed, t: 4m50s | PASS :white_check_mark: 
gcc.Debug <br/>-Dstatic=OFF | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2951/1//artifact/gcc.Debug_Dstatic_OFF.txt) | 1178 cases, 0 failed, t: 2m53s | PASS :white_check_mark: 
gcc.Debug <br/>-Dstatic=OFF -DBUILD_SHARED_LIBS=ON | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2951/1//artifact/gcc.Debug_Dstatic_OFF_DBUILD_SHARED_LIBS_ON.txt) | 1178 cases, 0 failed, t: 2m49s | PASS :white_check_mark: 
gcc.Debug, <br/>NINJA_BUILD=true | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2951/1//artifact/gcc.Debug_NINJA_BUILD_true.txt) | 1178 cases, 0 failed, t: 2m44s | PASS :white_check_mark: 
clang.Debug <br/>-Dunity=OFF -Dsan=address, <br/>PARALLEL_TESTS=false, <br/>DEBUGGER=false | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2951/1//artifact/clang.Debug_Dunity_OFF_Dsan_address_PARALLEL_TESTS_false_DEBUGGER_false.txt) | 1177 cases, 0 failed, t: 10m11s | PASS :white_check_mark: 
clang.Debug <br/>-Dunity=OFF -Dsan=undefined, <br/>PARALLEL_TESTS=false | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2951/1//artifact/clang.Debug_Dunity_OFF_Dsan_undefined_PARALLEL_TESTS_false.txt) | 1177 cases, 0 failed, t: 11m54s | PASS :white_check_mark:

thejohnfreeman commented:
Reading through this code for the first time, I can't decipher the meaning of the name of the member variable `pn`. Can we spell it out? There should at least be a comment. (I generally believe that the larger a variable's scope, the longer its name should be.)

thejohnfreeman commented on src/ripple/basics/base_uint.h:
```diff
@@ -126,25 +126,45 @@ class base_uint
+    template <class Container,
+              class = std::void_t
+              <
+                  decltype(std::declval<Container const>().size()),
```
Are these like a poor man's concepts? I'm not opposed to them, but just to confirm my understanding of C++, they aren't strictly necessary for correct compilation, are they?

HowardHinnant replied:
Yes, poor man's concepts.  They are necessary to prevent _incorrect_ compilation, such as accidentally constructing from a range of types that shouldn't be `memcpy`'d.

jwbusch replied:
Could these be pulled into their own type_trait? :) 
```
template<typename Container>
struct void_contiguous_container_t : std::void_t<
    decltype(std::declval<Container const>().size()),
    decltype(std::declval<Container const>().data()),
    typename Container::value_type
>{};

template<typename Container, typename Type = void>
using enable_if_trivially_copyable_t = std::enable_if_t<
    std::is_trivially_copyable<typename Container::value_type>::value, Type
>;
```

Then you can use it like... 

```
    template <class Container,
              class = void_contiguous_container_t<Container>,
              class = enable_if_trivially_copyable_t<Container>>
    explicit base_uint(Container const& c)
    {
        assert(c.size()*sizeof(typename Container::value_type) == size());
        std::memcpy(data_.data(), c.data(), size());
    }
```

I like giving names to custom traits, layering on the core type_traits. It's verbose, but doesn't require comments to explain it/less copy paste. The names I picked probably aren't the best, but that's the basic idea.

HowardHinnant replied:
We could.  I'd be more into it if we needed `void_contiguous_container_t` more places and more often.  With it just in `base_uint` I don't see a benefit to create this and move its definition off to a generic utility file.  However if people strongly feel that would make the code better, I don't mind doing it.

jwbusch replied:
My limit is usually ~1 duplication, but I know others are less bothered by it. It also doesn't need to be moved to a different file either, I usually just stick little utilities like these in an anonymous namespace at the top of the file after `#includes`. Then if it's used in multiple places, pull it out into a header.

It's mostly about readability and understanding the intent. I can read `contiguous_container_t` (or maybe `contiguous_container_concept` ...?) and immediately understand what the intent is, vs. reading those 3 lines and having to guess what the intent is. But it's just an idea 😃, I wouldn't block on it.

thejohnfreeman replied:
@HowardHinnant I was under the impression that the `std::is_trivially_copable` constraint is necessary for correct compilation, but that the member checks are not. I thought that template instantiation would just fail if those members were absent. Is that correct? I guess I should have been more clear that my comment was for the `std::void_t` expression only, not the `std::is_trivially_copyable` expression.

I weakly prefer @jwbusch's suggestion for a named trait, too, but I'm not going to block on it either.

HowardHinnant replied:
The constraints keep other constructors such as `base_uint (std::uint64_t b)` accidentally getting confused with the contiguous container constructor (and ditto for assignment).  I did not test that the constraints were necessary.  Just I've hit that bug so many times I've developed an aversion to "overly generic" signatures.

I'm testing a contiguous_container trait and will push it once I'm confident in it.

HowardHinnant replied:
And pushed.

thejohnfreeman replied:
I think the non-template always wins as more-specific (see [tie-breaker #4 under "Best viable function"](https://en.cppreference.com/w/cpp/language/overload_resolution)), and my [limited test](https://wandbox.org/permlink/mDXQk8DnKmmTOirY) suggests as much, but I'm totally fine with erring on the side of caution.

HowardHinnant commented:
@scottschurr @thejohnfreeman  I've pushed one more FOLD to rename the data member `pn` to `data_` and change its scope from protected to private.  Still ok?

scottschurr commented:
:+1:  Yup.  Looks like you're good to go.

thejohnfreeman commented on src/ripple/basics/base_uint.h:
```diff
@@ -37,6 +37,20 @@
+{
+
+template <class Container>
+using contiguous_container
```
How about naming this `is_contiguous_container` or `void_if_contiguous_container`?

HowardHinnant replied:
The `is_traits` typically evaluate to true or false.  And the `void_if` variant just seems wordy to me.

scottschurr commented on src/ripple/basics/base_uint.h:
```diff
@@ -37,6 +37,20 @@
+          decltype(std::declval<Container const>().size()),
+          decltype(std::declval<Container const>().data()),
+          typename Container::value_type
+      >;
```
Nice.  Just to prove to myself that this concept works, I tried this:
```
    {
        uint128 b;
        std::list<std::uint8_t> l (sizeof (b));
        b = l;
    }
```
The compiler diagnostic I got back (from clang) was:
```
no member named 'data' in 'std::__1::list<unsigned char...
```
Thanks for the education.

HowardHinnant commented on src/ripple/basics/base_uint.h:
```diff
@@ -287,7 +311,7 @@ class base_uint
         Hasher& h, base_uint const& a) noexcept
     {
         // Do not allow any endian transformations on this memory
-        h(a.pn.data (), sizeof(a.pn));
```
I believe the intent is to catch at compile-time use of non-cryptographic hashers such as those used to store objects in unordered containers.  Instead this just allows things like the SHA-family of hashers that specify endian.  That being said, this is not the right way to say that.  If we ported rippled to a big endian machine, this constraint would fire incorrectly.

We should either require big endian, or drop the constraint entirely.

For your second question, we want the SHA hasher to treat the data as an array of bytes, and not an array of `uint32_t`.  If the hasher saw an array of `uint32_t` it might be tempted to reverse the bytes on each `uint32_t` for endian purposes.  The fact that `base_uint` stores `uint32_t` instead of `uint8_t` should be invisible to all clients, even the hasher.

HowardHinnant commented on src/ripple/basics/base_uint.h:
```diff
@@ -126,25 +139,35 @@ class base_uint
     }
 
-    explicit base_uint (Blob const& vch)
+    explicit base_uint (std::uint64_t b)
```
I just tried to add it to this PR, but I'm not sure what to do about this one:

https://github.com/ripple/rippled/blob/develop/src/ripple/protocol/impl/UintTypes.cpp#L125

```
    static Currency const currency(0x5852500000000000);
```

HowardHinnant commented:
Ok, more polishing:

The constraints on the constructor and assignment weren't working.  For example `is_constructible` and `is_assignable` would give the wrong answer when the source was a `complex<T>`.  Fixed that and put a test in for it.

Changed many constructions using `0` to use `beast::zero` instead.  I likely didn't catch all of them.

Removed an errant constraint from `hash_append`.

Removed an unneeded type alias: `array_type`.

I don't think there's anything controversial here, but I'll hold off putting the passed tag back on to give reviewers a chance for another look.

HowardHinnant commented:
Squashed.

scottschurr commented:
Closing this pull request purely to reduce confusion.  The contents of this pull request now comprise one of the commits in https://github.com/ripple/rippled/pull/2922 which just passed review.  When https://github.com/ripple/rippled/pull/2922 is merged, then this pull request will also come along for the ride.
